contracts/dual_token/DualStakingToken.sol

```diff
@@ -0,0 +1,427 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+import {IService} from "./interfaces/IService.sol";
+import {IStaking} from "./interfaces/IStaking.sol";
+import {SafeTransferLib} from "../libraries/SafeTransferLib.sol";
+
+interface IActivityChecker {
+    /// @dev Locks activity checker.
+    function lock() external;
+
+    /// @dev Unlocks activity checker.
+    function unlock() external;
+}
+
+/// @dev Zero address.
+error ZeroAddress();
+
+/// @dev Zero value.
+error ZeroValue();
+
+/// @dev Only `owner` has a privilege, but the `sender` was provided.
+/// @param sender Sender address.
+/// @param owner Required sender address as an owner.
+error OwnerOnly(address sender, address owner);
+
+/// @dev Caught reentrancy violation.
+error ReentrancyGuard();
+
+struct StakerInfo {
+    // Staking token amount
+    uint256 stakingAmount;
+    // Cumulative reward
+    uint256 reward;
+    // Staker account address
+    address account;
+}
+
+/// @title DualStakingToken - Smart contract for dual token staking
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract DualStakingToken {
+    event OwnerUpdated(address indexed owner);
+    event StakingTokenParamsUpdated(uint256 stakingTokenAmount, uint256 rewardRatio);
+    event Deposit(address indexed sender, uint256 amount, uint256 balance, uint256 availableRewards);
+    event Withdraw(address indexed to, uint256 amount);
+
+    // Service registry address
+    address public immutable serviceRegistry;
+    // Staking token address (except for OLAS)
+    address public immutable stakingToken;
+    // Service staking instance address
+    address public immutable stakingInstance;
+
+    // Required staking token amount
+    uint256 public stakingTokenAmount;
+    // Staking token ratio to OLAS rewards in 1e18 form
+    uint256 public rewardRatio;
+    // Staking token contract balance
+    uint256 public balance;
+    // Staking token available rewards
+    uint256 public availableRewards;
+    // Owner address
+    address public owner;
+
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Mapping of staker account address => Staker info struct
+    mapping(uint256 => StakerInfo) public mapStakerInfos;
+    /// Mapping of staked OLAS service multisigs
+    mapping(address => bool) public mapMutisigs;
+
+    /// @dev DualStakingToken constructor.
+    /// @param _serviceRegistry Service registry address.
+    /// @param _stakingToken Staking token address.
+    /// @param _stakingInstance Service staking instance address.
+    /// @param _rewardRatio Staking token ratio to OLAS rewards in 1e18 form.
+    constructor(
+        address _serviceRegistry,
+        address _stakingToken,
```

Comment:
This is the "second" token? Should be made clear somehow in the docstring

Reply:
Renamed it eventually